atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -0,0 +1,428 @@
+       return runWithRetryOnHost(getRandomGoodHost().getHost(), f);
+    }
+
+    public <V, K extends Exception> V runWithRetryOnHost(InetSocketAddress specifiedHost, FunctionCheckedException<Cassandra.Client, V, K> f) throws K {
+        int numTries = 0;
+        while (true) {
+            CassandraClientPoolingContainer hostPool = currentPools.get(specifiedHost);
+
+            if (blacklistedHosts.containsKey(specifiedHost) || hostPool == null) {
+                log.warn("Randomly redirected a query intended for host %s because it was not currently a live member of the pool.", specifiedHost);
+                hostPool = getRandomGoodHost();
+            }
+
+            try {
+                return hostPool.runWithPooledResource(f);
+            } catch (Exception e) {
+                numTries++;
+                this.<K>handleException(numTries, hostPool.getHost(), e);
+            }
+        }
+    }
+
+    public <V, K extends Exception> V run(FunctionCheckedException<Cassandra.Client, V, K> f) throws K {
+        return runOnHost(getRandomGoodHost().getHost(), f);
+    }
+
+    public <V, K extends Exception> V runOnHost(InetSocketAddress specifiedHost, FunctionCheckedException<Cassandra.Client, V, K> f) throws K {
+        CassandraClientPoolingContainer hostPool = currentPools.get(specifiedHost);
+        return hostPool.runWithPooledResource(f);
+    }
+
+        @SuppressWarnings("unchecked")
+    private <K extends Exception> void handleException(int numTries, InetSocketAddress host, Exception e) throws K {
+        if (e instanceof CassandraClientFactory.ClientCreationFailedException
+                || e instanceof TTransportException
+                || e instanceof TimedOutException
+                || e instanceof SocketTimeoutException
+                || e instanceof UnavailableException
+                || e instanceof InsufficientConsistencyException) {
+            if (numTries >= MAX_TRIES) {
+                if (e instanceof TTransportException
+                        && e.getCause() != null
+                        && (e.getCause().getClass() == SocketException.class)) {
+                    String msg = "Error writing to Cassandra socket. Likely cause: Exceeded maximum thrift frame size; unlikely cause: network issues.";
+                    log.error("Tried to connect to cassandra " + numTries + " times. " + msg, e);
+                    e = new TTransportException(((TTransportException) e).getType(), msg, e);
+                } else {
+                    log.error("Tried to connect to cassandra " + numTries + " times.", e);
+                }
+                throw (K) e;
+            } else {
+                log.warn("Error occurred talking to cassandra. Attempt {} of {}.", numTries, MAX_TRIES, e);
+                if (e instanceof SocketTimeoutException
+                        || e instanceof UnavailableException) {
```

Comment:
need to handle `NoSuchElementException`

would really prefer `isConnectionException(Exception)` like  https://github.com/palantir/atlasdb/commit/ee691033cf37eeda20b9e390a0dffb57f131ce42#diff-43246c727757ea4bd900eaa8b5f27385R125